src/tools/pkcs15-tool.c

```diff
@@ -1675,6 +1682,34 @@ static void list_info(void)
 			count++;
 		}
 	}
+	printf("\n");
+	for (i = 0; i < SC_MAX_SUPPORTED_ALGORITHMS; i++) {
+		struct sc_supported_algo_info * sa = &p15card->tokeninfo->supported_algos[i];
+
+		if (sa->reference == 0 && sa->reference == 0 && sa->mechanism == 0
+				&& sa->parameters == NULL && sa->operations == 0 && sa->algo_ref == 0)
+					break;
+		printf("\t\t sc_supported_algo_info[%d]:\n", i);
```

Comment:
Could you use some more user friendly string?

Reply:
```diff
diff --git a/src/tools/pkcs15-tool.c b/src/tools/pkcs15-tool.c
index 5c5d7343..00f60fd9 100644
--- a/src/tools/pkcs15-tool.c
+++ b/src/tools/pkcs15-tool.c
@@ -1688,22 +1688,24 @@ static void list_info(void)
        for (i = 0; i < SC_MAX_SUPPORTED_ALGORITHMS; i++) {
                struct sc_supported_algo_info * sa = &p15card->tokeninfo->supported_algos[i];

-               if (sa->reference == 0 && sa->reference == 0 && sa->mechanism == 0
-                               && sa->parameters == NULL && sa->operations == 0 && sa->algo_ref == 0)
-                                       break;
-               printf("\t\t sc_supported_algo_info[%d]:\n", i);
-               printf("\t\t\t reference  : %d\n",sa->reference);
-               printf("\t\t\t mechanism  : %d\n",sa->mechanism);
-                       if (sa->parameters) {
-               /* TODO following may segfault parameters may point at garbage. */
-               //              if (sc_valid_oid(sa->parameters)) {
-               //                      printf("\t\t\t parameters:  %i", sa->parameters->value[0]);
-               //                      for (idx = 1; idx < SC_MAX_OBJECT_ID_OCTETS && sa->parameters->value[idx] != -1 ; idx++)
-               //                              printf(".%i", sa->parameters->value[idx]);
-                                       printf(" Parrameters found \n");
-               //      }
+               if (sa->reference)
+                       printf("\t\t\t reference  : %d\n",sa->reference);
+
+               if (sa->mechanism)
+                       printf("\t\t\t mechanism  : %d\n", sa->mechanism);
+
+               if (sa->operations)
+                       printf("\t\t\t operations : %d\n",sa->operations);
+
+               if (sa->parameters) {
+                       if (sc_valid_oid(sa->parameters)) {
+                               printf("\t\t\t parameters:  %i", sa->parameters->value[0]);
+                               for (idx = 1; idx < SC_MAX_OBJECT_ID_OCTETS && sa->parameters->value[idx] != -1 ; idx++)
+                                       printf(".%i", sa->parameters->value[idx]);
+                               printf(" Parrameters found \n");
+                       }
                }
-               printf("\t\t\t operations : %d\n",sa->operations);
+
                if (sc_valid_oid((const struct sc_object_id*)&sa->algo_id)) {
```

Reply:
I used the term `sc_supported_algo_info` and the variable names from the OpenSC structure in the printf. This code is not complete, as there is still the outstanding SEGFAULT, most likely caused because the structure is copied, and `parameters` is a pointer to an OID, and not accounted for during the copy.

Reply:
the segfault issue is resolved, right?

Reply:
Correct. 
./pkcs15.h has the `struct sc_supported_algo_info` inline:
```
 566         struct sc_supported_algo_info supported_algos[SC_MAX_SUPPORTED_ALGORITHMS];
 567 } sc_pkcs15_tokeninfo_t;
```
`opensc.h` has the two `sc_object_id`  inline:
```
 223 struct sc_supported_algo_info {
 224         unsigned int reference;
 225         unsigned int mechanism;
 226         struct sc_object_id parameters; /* OID for ECC */
 227         unsigned int operations;
 228         struct sc_object_id algo_id;
 229         unsigned int algo_ref;
 230 };
```
The `p15card->tokeninfo` is allocated in `pkcs15.c`:
```
702         p15card->tokeninfo = calloc(1, sizeof(struct sc_pkcs15_tokeninfo));
```
So even if copied, there are no pointers.

It should be on a TODO list to fix asn.1 to read an array of entries using realloc and save a count of entries.  Anything using SC_MAX_SUPPORTED_ALGORITHMS would use the pointer and count.  And  provide copy and free routines if needed.